src/program.cpp

```diff
@@ -760,176 +419,66 @@ void program::debug_print(instruction_ref ins) const
         }
     });
 }
-void program::debug_print(const std::vector<instruction_ref>& inss) const
-{
-    for(auto ins : inss)
-        debug_print(ins);
-    std::cout << std::endl;
-}
 
-static std::string enclose_name(const std::string& name)
+void program::print_graph(std::ostream& os, bool brief) const
 {
-    return '"' + replace_string(name, "\"", "\\\"") + '"';
+    const auto* mm = this->get_main_module();
+    mm->print_graph(os, brief);
 }
 
-void program::print_graph(std::ostream& os, bool brief) const
+void program::print_cpp(std::ostream& os) const
 {
-    os << "digraph {" << std::endl;
-    os << "\trankdir=LR;" << std::endl;
-    print_program(*this, [&](auto ins, const auto& names) {
-        std::string label;
-        if(brief)
-            label = ins->name();
-        else
-            label = to_string(ins->get_operator());
-        os << "\t" << enclose_name(names.at(ins)) << "[label=" << enclose_name(label) << "]";
-        os << ";" << std::endl;
-        if(!ins->inputs().empty())
-        {
-            for(auto&& arg : ins->inputs())
-            {
-                os << "\t" << enclose_name(names.at(arg)) << " -> " << enclose_name(names.at(ins));
-                if(not brief)
-                    os << "[label=" << enclose_name(to_string(ins->get_shape())) << "]";
-                os << ";" << std::endl;
-            }
-        }
-    });
-    os << "}" << std::endl;
+    os << "migraphx::program p;" << std::endl;
+    const auto* mm = this->get_main_module();
+    mm->print_cpp(os);
 }
 
-static std::string cpp_var_name(const std::string& name)
+void program::dry_run(std::unordered_map<std::string, argument> params) const
 {
-    return "m" + replace_string(name, "@", "x");
+    auto& ctx = this->impl->ctx;
+    generic_eval(*this, ctx, std::move(params), [](auto&&...) { return argument{}; });
 }
 
-static std::string cpp_op_var(const std::string& name, instruction_ref ins)
+void program::annotate(std::ostream& os, std::function<void(instruction_ref)> a) const
 {
-    return replace_string(name, "@", ins->name());
+    const auto* mm = this->get_main_module();
+    std::cout << mm->name() << ":" << std::endl;
+    mm->annotate(os, std::move(a));
 }
 
-static void print_op_attributes(std::ostream& os, const std::string& name, const operation& op)
+module* program::get_main_module()
 {
-    std::string x = to_string(op);
-    if(contains(x, "["))
+    if(!contains(impl->modules, "main"))
```

Comment:
There is no need to search twice, just do `return &impl->modules["main"]`.

Reply:
fixed

Reply:
This condition can be removed.

Reply:
removed